### ISSUE TITLE
Deadlock in Deallocators.

### ISSUE THREAD

**Issue by pokowaka** [deeplearning4j/rl4j#77]
#### Issue Description
- After a few hours of training the system deadlocks.

Log:

04:44:11.604 [main] INFO  o.d.rl4j.learning.sync.SyncLearning - Epoch: 160929, reward: 7.0
04:44:11.678 [main] INFO  o.d.rl4j.learning.sync.SyncLearning - Epoch: 160930, reward: 0.0
04:44:11.711 [main] INFO  o.d.o.l.ScoreIterationListener - Score at iteration 2444400 is 3.9150946625005944E-7
04:44:11.785 [main] INFO  o.d.rl4j.learning.sync.SyncLearning - Epoch: 160931, reward: 3.0
04:44:11.819 [main] DEBUG o.d.optimize.solvers.BaseOptimizer - Hit termination condition on iteration 0: score=5.581014911034512E-8, oldScore=5.615988572428346E-8, condition=org.deeplearning4j.optimize.terminations.EpsTermination@41991619
04:44:11.893 [main] INFO  o.d.rl4j.learning.sync.SyncLearning - Epoch: 160932, reward: 3.0
04:44:11.973 [main] INFO  o.d.rl4j.learning.sync.SyncLearning - Epoch: 160933, reward: 0.0
<=========----> 75% EXECUTING [8h 25m 41s]

Trace:

/System/Library/Frameworks/JavaVM.framework/Versions/Current/Commands/jsadebugd 26720
Attaching to process ID 26720 and starting RMI services, please wait...
Debugger attached and RMI services started.

jdb -connect sun.jvm.hotspot.jdi.SADebugServerAttachingConnector:debugServerName=localhost
Initializing jdb ...
> threads
Group system:
  (java.lang.ref.Reference$ReferenceHandler)0x5                                         Reference Handler                cond. waiting
  (java.lang.ref.Finalizer$FinalizerThread)0x4                                          Finalizer                        cond. waiting
  (java.lang.Thread)0x3                                                                 Signal Dispatcher                running
Group main:
  (java.lang.Thread)0x6                                                                 main                             running
  (org.nd4j.linalg.memory.provider.BasicWorkspaceManager$WorkspaceDeallocatorThread)0x2 Workspace deallocator thread     cond. waiting
  (org.bytedeco.javacpp.Pointer$DeallocatorThread)0x1                                   JavaCPP Deallocator              cond. waiting
  (org.nd4j.rng.deallocator.NativeRandomDeallocator$DeallocatorThread)0x0               NativeRandomDeallocator thread 0 cond. waiting

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version: 0.9.2-SNAPSHOT
* platform information: mac os sierria
* CUDA version, if used, none running on cpu.

#### Contributing

I didn't start the JVM with debugging enabled. Will try to get more.

**Comment by raver119** [comment]
All those threads should be daemonic.
Please post full threads dump.

**Comment by pokowaka** [comment]
Full dump:

jstack -l 26720
2017-10-13 09:47:36
Full thread dump Java HotSpot(TM) 64-Bit Server VM (25.144-b01 mixed mode):

"Attach Listener" #14 daemon prio=9 os_prio=31 tid=0x00007f902bfc6000 nid=0x130f waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"NativeRandomDeallocator thread 0" #13 daemon prio=5 os_prio=31 tid=0x00007f902bb9f000 nid=0x5e03 in Object.wait() [0x000070000bf43000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c0209c28> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x00000006c0209c28> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at org.nd4j.rng.deallocator.NativeRandomDeallocator$DeallocatorThread.run(NativeRandomDeallocator.java:77)

   Locked ownable synchronizers:
	- None

"JavaCPP Deallocator" #12 daemon prio=10 os_prio=31 tid=0x00007f902d26e000 nid=0x5903 in Object.wait() [0x000070000bcb7000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x00000006c0005388> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at org.bytedeco.javacpp.Pointer$DeallocatorThread.run(Pointer.java:299)

   Locked ownable synchronizers:
	- None

"Workspace deallocator thread" #11 daemon prio=5 os_prio=31 tid=0x00007f902d24c000 nid=0x5703 in Object.wait() [0x000070000bbb4000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	- waiting on <0x00000006c0005538> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x00000006c0005538> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at org.nd4j.linalg.memory.provider.BasicWorkspaceManager$WorkspaceDeallocatorThread.run(BasicWorkspaceManager.java:246)

   Locked ownable synchronizers:
	- None

"Service Thread" #9 daemon prio=9 os_prio=31 tid=0x00007f902b820800 nid=0x5303 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C1 CompilerThread3" #8 daemon prio=9 os_prio=31 tid=0x00007f902d040800 nid=0x5103 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C2 CompilerThread2" #7 daemon prio=9 os_prio=31 tid=0x00007f902e02e800 nid=0x4f03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C2 CompilerThread1" #6 daemon prio=9 os_prio=31 tid=0x00007f902d067000 nid=0x4d03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"C2 CompilerThread0" #5 daemon prio=9 os_prio=31 tid=0x00007f902b815000 nid=0x4b03 waiting on condition [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"Signal Dispatcher" #4 daemon prio=9 os_prio=31 tid=0x00007f902b803000 nid=0x4903 runnable [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
	- None

"Finalizer" #3 daemon prio=8 os_prio=31 tid=0x00007f902e00a000 nid=0x3903 in Object.wait() [0x000070000b39c000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:143)
	- locked <0x00000006c001b210> (a java.lang.ref.ReferenceQueue$Lock)
	at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:164)
	at java.lang.ref.Finalizer$FinalizerThread.run(Finalizer.java:209)

   Locked ownable synchronizers:
	- None

"Reference Handler" #2 daemon prio=10 os_prio=31 tid=0x00007f902e007000 nid=0x3703 in Object.wait() [0x000070000b299000]
   java.lang.Thread.State: WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at java.lang.ref.Reference.tryHandlePending(Reference.java:191)
	- locked <0x00000006c0002bf8> (a java.lang.ref.Reference$Lock)
	at java.lang.ref.Reference$ReferenceHandler.run(Reference.java:153)

   Locked ownable synchronizers:
	- None

"main" #1 prio=5 os_prio=31 tid=0x00007f902d008800 nid=0x1c03 runnable [0x000070000a87a000]
   java.lang.Thread.State: RUNNABLE
	at org.nd4j.nativeblas.Nd4jCpu$NativeOps.execScalarFloat(Native Method)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.invoke(NativeOpExecutioner.java:566)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:596)
	at org.nd4j.linalg.cpu.nativecpu.ops.NativeOpExecutioner.exec(NativeOpExecutioner.java:107)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.doRowWise(BaseNDArray.java:2412)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.addiRowVector(BaseNDArray.java:2791)
	at org.deeplearning4j.nn.layers.BaseLayer.preOutput(BaseLayer.java:336)
	at org.deeplearning4j.nn.layers.BaseLayer.backpropGradient(BaseLayer.java:73)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.calcBackpropGradients(MultiLayerNetwork.java:1354)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.backprop(MultiLayerNetwork.java:1268)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.computeGradientAndScore(MultiLayerNetwork.java:2235)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:173)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:60)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:53)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1776)
	at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.fit(MultiLayerNetwork.java:1725)
	at org.deeplearning4j.rl4j.network.dqn.DQN.fit(DQN.java:46)
	at org.deeplearning4j.rl4j.learning.sync.qlearning.discrete.QLearningDiscrete.trainStep(QLearningDiscrete.java:184)
	at org.deeplearning4j.rl4j.learning.sync.qlearning.QLearning.trainEpoch(QLearning.java:102)
	at org.deeplearning4j.rl4j.learning.sync.SyncLearning.train(SyncLearning.java:44)
	at org.pokowaka.hanabi.Hanabi.train(Hanabi.java:42)
	at org.pokowaka.hanabi.Hanabi.main(Hanabi.java:71)

   Locked ownable synchronizers:
	- None

"VM Thread" os_prio=31 tid=0x00007f902d066000 nid=0x3503 runnable

"GC task thread#0 (ParallelGC)" os_prio=31 tid=0x00007f902d014800 nid=0x2503 runnable

"GC task thread#1 (ParallelGC)" os_prio=31 tid=0x00007f902d015000 nid=0x2703 runnable

"GC task thread#2 (ParallelGC)" os_prio=31 tid=0x00007f902d015800 nid=0x2903 runnable

"GC task thread#3 (ParallelGC)" os_prio=31 tid=0x00007f902d016000 nid=0x2b03 runnable

"GC task thread#4 (ParallelGC)" os_prio=31 tid=0x00007f902e000800 nid=0x2d03 runnable

"GC task thread#5 (ParallelGC)" os_prio=31 tid=0x00007f902e001800 nid=0x2f03 runnable

"GC task thread#6 (ParallelGC)" os_prio=31 tid=0x00007f902e002000 nid=0x3103 runnable

"GC task thread#7 (ParallelGC)" os_prio=31 tid=0x00007f902e002800 nid=0x3303 runnable

"VM Periodic Task Thread" os_prio=31 tid=0x00007f902d041000 nid=0x5503 waiting on condition

JNI global references: 241

**Comment by pokowaka** [comment]
(lldb) thread list
Process 26720 stopped
* thread #1: tid = 0x1310efb, 0x00007fffea63b34a libsystem_kernel.dylib`mach_msg_trap + 10, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  thread #2: tid = 0x1310f01, 0x00007fffea642f46 libsystem_kernel.dylib`__semwait_signal + 10
  thread #3: tid = 0x1310f05, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #4: tid = 0x1310f06, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #5: tid = 0x1310f07, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #6: tid = 0x1310f08, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #7: tid = 0x1310f09, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #8: tid = 0x1310f0a, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #9: tid = 0x1310f0b, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #10: tid = 0x1310f0c, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #11: tid = 0x1310f0d, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #12: tid = 0x1310f0e, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #13: tid = 0x1310f0f, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: Reference Handler'
  thread #14: tid = 0x1310f10, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: Finalizer'
  thread #15: tid = 0x1310f11, 0x00007fffea63b386 libsystem_kernel.dylib`semaphore_wait_trap + 10, name = 'Java: Signal Dispatcher'
  thread #16: tid = 0x1310f12, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: C2 CompilerThread0'
  thread #17: tid = 0x1310f13, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: C2 CompilerThread1'
  thread #18: tid = 0x1310f14, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: C2 CompilerThread2'
  thread #19: tid = 0x1310f15, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: C1 CompilerThread3'
  thread #20: tid = 0x1310f16, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: Service Thread'
  thread #21: tid = 0x1310f17, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #22: tid = 0x1310f19, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: Workspace deallocator thread'
  thread #23: tid = 0x1310f1b, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: JavaCPP Deallocator'
  thread #24: tid = 0x1310f28, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #25: tid = 0x1310f29, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #26: tid = 0x1310f2a, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #27: tid = 0x1310f2d, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10, name = 'Java: NativeRandomDeallocator thread 0'
  thread #28: tid = 0x1310f47, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #29: tid = 0x1310f48, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #30: tid = 0x1310f49, 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
  thread #31: tid = 0x135cb8d, 0x00007fffea64344e libsystem_kernel.dylib`__workq_kernreturn + 10
  thread #32: tid = 0x13625cf, 0x00007fffea6420c2 libsystem_kernel.dylib`__accept + 10, name = 'Java: Attach Listener'

**Comment by pokowaka** [comment]
Thread 28 & 30?

(lldb) thread backtrace all
  thread #1, queue = 'com.apple.main-thread'
    frame #0: 0x00007fffea63b34a libsystem_kernel.dylib`mach_msg_trap + 10
    frame #1: 0x00007fffea63a848 libsystem_kernel.dylib`mach_msg + 232
    frame #2: 0x00007fffd4d8d434 CoreFoundation`__CFRunLoopServiceMachPort + 212
    frame #3: 0x00007fffd4d8c8c1 CoreFoundation`__CFRunLoopRun + 1361
    frame #4: 0x00007fffd4d8c114 CoreFoundation`CFRunLoopRunSpecific + 420
    frame #5: 0x0000000102dc2465 java`CreateExecutionEnvironment + 871
    frame #6: 0x0000000102dbe18c java`JLI_Launch + 1952
    frame #7: 0x0000000102dc44c2 java`main + 101
    frame #8: 0x0000000102dbd9e4 java`start + 52

  thread #2
    frame #0: 0x00007fffea642f46 libsystem_kernel.dylib`__semwait_signal + 10
    frame #1: 0x00007fffea72f6e5 libsystem_pthread.dylib`pthread_join + 425
    frame #2: 0x0000000102dc1a22 java`ContinueInNewThread0 + 102
    frame #3: 0x0000000102dbef53 java`ContinueInNewThread + 201
    frame #4: 0x0000000102dc17f7 java`JVMInit + 315
    frame #5: 0x0000000102dbecd1 java`JLI_Launch + 4837
    frame #6: 0x0000000102dc44c2 java`main + 101
    frame #7: 0x0000000102dc20f7 java`apple_main + 92
    frame #8: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #9: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #10: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #3
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000012322346b libnd4jcpu.dylib`gomp_sem_wait(sem=0x00007f902c8fa518) at sem.c:71
    frame #3: 0x00000001232235f9 libnd4jcpu.dylib`gomp_barrier_wait_end(bar=0x00007f902c8fa460, state=<unavailable>) at bar.c:85
    frame #4: 0x0000000123221e2a libnd4jcpu.dylib`gomp_team_start(fn=(libnd4jcpu.dylib`void functions::scalar::ScalarTransform<float>::transform<simdOps::Add<float> >(float*, int*, float*, float*, int*, float*, int*, int, int*, long long*, int*, long long*) (._omp_fn.0)), data=0x000070000a87ae20, nthreads=4, flags=<unavailable>, team=0x00007f9030bd0000) at team.c:805
    frame #5: 0x000000012321e7ea libnd4jcpu.dylib`GOMP_parallel(fn=(libnd4jcpu.dylib`void functions::scalar::ScalarTransform<float>::transform<simdOps::Add<float> >(float*, int*, float*, float*, int*, float*, int*, int, int*, long long*, int*, long long*) (._omp_fn.0)), data=0x000070000a87ae20, num_threads=4, flags=3) at parallel.c:167
    frame #6: 0x00000001227aef3f libnd4jcpu.dylib`functions::scalar::ScalarTransform<float>::transform(int, float*, int*, float*, float*, int*, float*, int*, int, int*, long long*, int*, long long*) + 1519
    frame #7: 0x0000000122742e5f libnd4jcpu.dylib`NativeOps::execScalarFloat(void**, int, float*, int*, float*, int*, float*, float*, int*, int) + 95
    frame #8: 0x0000000123650c1b libjnind4jcpu.dylib`Java_org_nd4j_nativeblas_Nd4jCpu_00024NativeOps_execScalarFloat__Lorg_bytedeco_javacpp_PointerPointer_2ILorg_bytedeco_javacpp_FloatPointer_2Lorg_bytedeco_javacpp_IntPointer_2Lorg_bytedeco_javacpp_FloatPointer_2Lorg_bytedeco_javacpp_IntPointer_2Lorg_bytedeco_javacpp_FloatPointer_2Lorg_bytedeco_javacpp_FloatPointer_2Lorg_bytedeco_javacpp_IntPointer_2I + 715
    frame #9: 0x00000001063d8cea

  thread #4
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x0000000104e761be libjvm.dylib`GCTaskManager::get_task(unsigned int) + 56
    frame #7: 0x0000000104e7728b libjvm.dylib`GCTaskThread::run() + 349
    frame #8: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #9: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #10: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #11: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #5
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x0000000104e761be libjvm.dylib`GCTaskManager::get_task(unsigned int) + 56
    frame #7: 0x0000000104e7728b libjvm.dylib`GCTaskThread::run() + 349
    frame #8: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #9: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #10: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #11: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #6
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x0000000104e761be libjvm.dylib`GCTaskManager::get_task(unsigned int) + 56
    frame #7: 0x0000000104e7728b libjvm.dylib`GCTaskThread::run() + 349
    frame #8: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #9: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #10: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #11: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #7
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x0000000104e761be libjvm.dylib`GCTaskManager::get_task(unsigned int) + 56
    frame #7: 0x0000000104e7728b libjvm.dylib`GCTaskThread::run() + 349
    frame #8: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #9: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #10: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #11: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #8
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x0000000104e761be libjvm.dylib`GCTaskManager::get_task(unsigned int) + 56
    frame #7: 0x0000000104e7728b libjvm.dylib`GCTaskThread::run() + 349
    frame #8: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #9: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #10: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #11: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #9
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x0000000104e761be libjvm.dylib`GCTaskManager::get_task(unsigned int) + 56
    frame #7: 0x0000000104e7728b libjvm.dylib`GCTaskThread::run() + 349
    frame #8: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #9: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #10: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #11: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #10
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x0000000104e761be libjvm.dylib`GCTaskManager::get_task(unsigned int) + 56
    frame #7: 0x0000000104e7728b libjvm.dylib`GCTaskThread::run() + 349
    frame #8: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #9: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #10: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #11: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #11
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x0000000104e761be libjvm.dylib`GCTaskManager::get_task(unsigned int) + 56
    frame #7: 0x0000000104e7728b libjvm.dylib`GCTaskThread::run() + 349
    frame #8: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #9: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #10: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #11: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #12
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x0000000105088774 libjvm.dylib`os::PlatformEvent::park(long) + 404
    frame #3: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #4: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #5: 0x00000001051b83b6 libjvm.dylib`VMThread::loop() + 444
    frame #6: 0x00000001051b7e41 libjvm.dylib`VMThread::run() + 121
    frame #7: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #8: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #9: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #10: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #13, name = 'Java: Reference Handler'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010507f3ba libjvm.dylib`ObjectMonitor::wait(long, bool, Thread*) + 748
    frame #4: 0x0000000105138a00 libjvm.dylib`ObjectSynchronizer::wait(Handle, long, Thread*) + 202
    frame #5: 0x0000000104f501c3 libjvm.dylib`JVM_MonitorWait + 156
    frame #6: 0x0000000105e20928

  thread #14, name = 'Java: Finalizer'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010507f3ba libjvm.dylib`ObjectMonitor::wait(long, bool, Thread*) + 748
    frame #4: 0x0000000105138a00 libjvm.dylib`ObjectSynchronizer::wait(Handle, long, Thread*) + 202
    frame #5: 0x0000000104f501c3 libjvm.dylib`JVM_MonitorWait + 156
    frame #6: 0x0000000105e20928

  thread #15, name = 'Java: Signal Dispatcher'
    frame #0: 0x00007fffea63b386 libsystem_kernel.dylib`semaphore_wait_trap + 10
    frame #1: 0x0000000105089ddc libjvm.dylib`check_pending_signals(bool) + 128
    frame #2: 0x0000000105086b77 libjvm.dylib`signal_thread_entry(JavaThread*, Thread*) + 667
    frame #3: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #4: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #5: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #6: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #7: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #8: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #16, name = 'Java: C2 CompilerThread0'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x0000000105088774 libjvm.dylib`os::PlatformEvent::park(long) + 404
    frame #3: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #4: 0x0000000105065404 libjvm.dylib`Monitor::wait(bool, long, bool) + 222
    frame #5: 0x0000000104db93c0 libjvm.dylib`CompileQueue::get() + 122
    frame #6: 0x0000000104db95e2 libjvm.dylib`CompileBroker::compiler_thread_loop() + 376
    frame #7: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #8: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #9: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #10: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #11: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #12: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #17, name = 'Java: C2 CompilerThread1'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x0000000105088774 libjvm.dylib`os::PlatformEvent::park(long) + 404
    frame #3: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #4: 0x0000000105065404 libjvm.dylib`Monitor::wait(bool, long, bool) + 222
    frame #5: 0x0000000104db93c0 libjvm.dylib`CompileQueue::get() + 122
    frame #6: 0x0000000104db95e2 libjvm.dylib`CompileBroker::compiler_thread_loop() + 376
    frame #7: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #8: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #9: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #10: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #11: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #12: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #18, name = 'Java: C2 CompilerThread2'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x0000000105088774 libjvm.dylib`os::PlatformEvent::park(long) + 404
    frame #3: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #4: 0x0000000105065404 libjvm.dylib`Monitor::wait(bool, long, bool) + 222
    frame #5: 0x0000000104db93c0 libjvm.dylib`CompileQueue::get() + 122
    frame #6: 0x0000000104db95e2 libjvm.dylib`CompileBroker::compiler_thread_loop() + 376
    frame #7: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #8: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #9: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #10: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #11: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #12: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #19, name = 'Java: C1 CompilerThread3'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x0000000105088774 libjvm.dylib`os::PlatformEvent::park(long) + 404
    frame #3: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #4: 0x0000000105065404 libjvm.dylib`Monitor::wait(bool, long, bool) + 222
    frame #5: 0x0000000104db93c0 libjvm.dylib`CompileQueue::get() + 122
    frame #6: 0x0000000104db95e2 libjvm.dylib`CompileBroker::compiler_thread_loop() + 376
    frame #7: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #8: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #9: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #10: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #11: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #12: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #20, name = 'Java: Service Thread'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010506495e libjvm.dylib`ParkCommon(ParkEvent*, long) + 42
    frame #4: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #5: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #6: 0x00000001050e38dc libjvm.dylib`ServiceThread::service_thread_entry(JavaThread*, Thread*) + 132
    frame #7: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #8: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #9: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #10: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #11: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #12: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #21
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x0000000105088774 libjvm.dylib`os::PlatformEvent::park(long) + 404
    frame #3: 0x0000000105065234 libjvm.dylib`Monitor::IWait(Thread*, long) + 168
    frame #4: 0x000000010506549d libjvm.dylib`Monitor::wait(bool, long, bool) + 375
    frame #5: 0x000000010516c590 libjvm.dylib`WatcherThread::sleep() const + 126
    frame #6: 0x000000010516d415 libjvm.dylib`WatcherThread::run() + 243
    frame #7: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #8: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #9: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #10: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #22, name = 'Java: Workspace deallocator thread'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010507f3ba libjvm.dylib`ObjectMonitor::wait(long, bool, Thread*) + 748
    frame #4: 0x0000000105138a00 libjvm.dylib`ObjectSynchronizer::wait(Handle, long, Thread*) + 202
    frame #5: 0x0000000104f501c3 libjvm.dylib`JVM_MonitorWait + 156
    frame #6: 0x0000000105bbe9f4
    frame #7: 0x0000000105e26e34
    frame #8: 0x0000000105ba77a7
    frame #9: 0x0000000104eee70a libjvm.dylib`JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*) + 1710
    frame #10: 0x0000000104eeeeae libjvm.dylib`JavaCalls::call_virtual(JavaValue*, KlassHandle, Symbol*, Symbol*, JavaCallArguments*, Thread*) + 356
    frame #11: 0x0000000104eef05a libjvm.dylib`JavaCalls::call_virtual(JavaValue*, Handle, KlassHandle, Symbol*, Symbol*, Thread*) + 74
    frame #12: 0x0000000104f497b1 libjvm.dylib`thread_entry(JavaThread*, Thread*) + 124
    frame #13: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #14: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #15: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #16: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #17: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #18: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

* thread #23, name = 'Java: JavaCPP Deallocator'
  * frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010507f3ba libjvm.dylib`ObjectMonitor::wait(long, bool, Thread*) + 748
    frame #4: 0x0000000105138a00 libjvm.dylib`ObjectSynchronizer::wait(Handle, long, Thread*) + 202
    frame #5: 0x0000000104f501c3 libjvm.dylib`JVM_MonitorWait + 156
    frame #6: 0x0000000105e20928

  thread #24
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x00000001249b00db libopenblas.dylib`blas_thread_server + 299
    frame #3: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #4: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #5: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #25
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x00000001249b00db libopenblas.dylib`blas_thread_server + 299
    frame #3: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #4: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #5: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #26
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x00000001249b00db libopenblas.dylib`blas_thread_server + 299
    frame #3: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #4: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #5: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #27, name = 'Java: NativeRandomDeallocator thread 0'
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000010508788a libjvm.dylib`os::PlatformEvent::park() + 192
    frame #3: 0x000000010507f3ba libjvm.dylib`ObjectMonitor::wait(long, bool, Thread*) + 748
    frame #4: 0x0000000105138a00 libjvm.dylib`ObjectSynchronizer::wait(Handle, long, Thread*) + 202
    frame #5: 0x0000000104f501c3 libjvm.dylib`JVM_MonitorWait + 156
    frame #6: 0x0000000105bbe9f4
    frame #7: 0x0000000105e26e34
    frame #8: 0x0000000105ba77a7
    frame #9: 0x0000000104eee70a libjvm.dylib`JavaCalls::call_helper(JavaValue*, methodHandle*, JavaCallArguments*, Thread*) + 1710
    frame #10: 0x0000000104eeeeae libjvm.dylib`JavaCalls::call_virtual(JavaValue*, KlassHandle, Symbol*, Symbol*, JavaCallArguments*, Thread*) + 356
    frame #11: 0x0000000104eef05a libjvm.dylib`JavaCalls::call_virtual(JavaValue*, Handle, KlassHandle, Symbol*, Symbol*, Thread*) + 74
    frame #12: 0x0000000104f497b1 libjvm.dylib`thread_entry(JavaThread*, Thread*) + 124
    frame #13: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #14: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #15: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #16: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #17: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #18: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #28
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000012322346b libnd4jcpu.dylib`gomp_sem_wait(sem=0x00007f902c8fa4a0) at sem.c:71
    frame #3: 0x000000012322361e libnd4jcpu.dylib`gomp_barrier_wait_end(bar=0x00007f902c8fa460, state=<unavailable>) at bar.c:92
    frame #4: 0x00000001232217e8 libnd4jcpu.dylib`gomp_thread_start(xdata=<unavailable>) at team.c:122
    frame #5: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #6: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #7: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #29
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000012322346b libnd4jcpu.dylib`gomp_sem_wait(sem=0x00007f9030bd0128) at sem.c:71
    frame #3: 0x000000012322374d libnd4jcpu.dylib`gomp_team_barrier_wait_end(bar=0x00007f9030bd00e8, state=0) at bar.c:151
    frame #4: 0x0000000123221815 libnd4jcpu.dylib`gomp_thread_start [inlined] gomp_team_barrier_wait_final(bar=<unavailable>) at bar.h:104
    frame #5: 0x0000000123221810 libnd4jcpu.dylib`gomp_thread_start(xdata=<unavailable>)
    frame #6: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #7: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #8: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #30
    frame #0: 0x00007fffea642bf2 libsystem_kernel.dylib`__psynch_cvwait + 10
    frame #1: 0x00007fffea72e7fa libsystem_pthread.dylib`_pthread_cond_wait + 712
    frame #2: 0x000000012322346b libnd4jcpu.dylib`gomp_sem_wait(sem=0x00007f902c8fa4a0) at sem.c:71
    frame #3: 0x000000012322361e libnd4jcpu.dylib`gomp_barrier_wait_end(bar=0x00007f902c8fa460, state=<unavailable>) at bar.c:92
    frame #4: 0x00000001232217e8 libnd4jcpu.dylib`gomp_thread_start(xdata=<unavailable>) at team.c:122
    frame #5: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #6: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #7: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13

  thread #31
    frame #0: 0x00007fffea64344e libsystem_kernel.dylib`__workq_kernreturn + 10
    frame #1: 0x00007fffea72d48e libsystem_pthread.dylib`_pthread_wqthread + 1023
    frame #2: 0x00007fffea72d07d libsystem_pthread.dylib`start_wqthread + 13

  thread #32, name = 'Java: Attach Listener'
    frame #0: 0x00007fffea6420c2 libsystem_kernel.dylib`__accept + 10
    frame #1: 0x0000000104cccfbe libjvm.dylib`BsdAttachListener::dequeue() + 122
    frame #2: 0x0000000104ccd023 libjvm.dylib`AttachListener::dequeue() + 69
    frame #3: 0x0000000104ccbf1f libjvm.dylib`attach_listener_thread_entry(JavaThread*, Thread*) + 362
    frame #4: 0x000000010516c0b3 libjvm.dylib`JavaThread::thread_main_inner() + 155
    frame #5: 0x000000010516d7a0 libjvm.dylib`JavaThread::run() + 450
    frame #6: 0x000000010508baee libjvm.dylib`java_start(Thread*) + 246
    frame #7: 0x00007fffea72d93b libsystem_pthread.dylib`_pthread_body + 180
    frame #8: 0x00007fffea72d887 libsystem_pthread.dylib`_pthread_start + 286
    frame #9: 0x00007fffea72d08d libsystem_pthread.dylib`thread_start + 13
(lldb)

**Comment by raver119** [comment]
Are you on macos x with skylake cpu by any chance?
We’ve heard about issues with cpus there, and had one user with similar behavior.

**Comment by pokowaka** [comment]
I'm running on:

    Hardware Overview:

      Model Name: MacBook Pro
      Model Identifier: MacBookPro11,4
      Processor Name: Intel Core i7
      Processor Speed: 2.2 GHz
      Number of Processors: 1
      Total Number of Cores: 4
      L2 Cache (per Core): 256 KB
      L3 Cache: 6 MB
      Memory: 16 GB
      Boot ROM Version: MBP114.0172.B25
      SMC Version (system): 2.29f24

**Comment by raver119** [comment]
No, you shouldn’t be affected then.

try to set OMP_NUM_THREADS=1 env variable, and check if that helps.

> 13 окт. 2017 г., в 20:22, Erwin Jansen <notifications@github.com> написал(а):
> 
> MacBookPro11,4

**Comment by saudet** [comment]
Which MDP does this happen with?

**Comment by saudet** [comment]
Any updates on this? Does setting OMP_NUM_THREADS=1 help? If so, this sounds like a known issue with OpenMP on Mac...

**Comment by akhodakivskiy** [comment]
I'm also seeing hanging upon program termination in 0.9.1, here is my thread dump:

https://gist.github.com/akhodakivskiy/07922f59a3c669d77179e3394dd4b3c7

**Comment by saudet** [comment]
Yeah, I also see some daemon thread not terminating for some reason sometimes, but anyway, if it's happening during training, that's another issue, I think?

**Comment by edschwarz** [comment]
I started seeing this using 0.9.1 on my MacBook Pro when I switched my maven dependency from rl4j-example to rl4j-core. 

I build a shaded jar, and the rl4j-core dependency seemed to work fine and result in a much smaller uber jar.  At runtime, it did not deadlock, but this started showing up in the logs: 

DEBUG o.d.optimize.solvers.BaseOptimizer - Hit termination condition on iteration 0: score=0.003649275140943064, oldScore=0.0036493099597728687, condition=org.deeplearning4j.optimize.terminations.EpsTermination@42313f13

I can reliably stop/start these messages by switching the dependency. My Mac is 
Hardware Overview:

  Model Name:	MacBook Pro
  Model Identifier:	MacBookPro11,4
  Processor Name:	Intel Core i7
  Processor Speed:	2.8 GHz
  Number of Processors:	1
  Total Number of Cores:	4
  L2 Cache (per Core):	256 KB
  L3 Cache:	6 MB
  Memory:	16 GB
  Boot ROM Version:	MBP114.0177.B00
  SMC Version (system):	2.29f24
  Serial Number (system):	C02V90CRG8WL
  Hardware UUID:	EA1556BF-9EEA-5DF3-B752-4CF3B0967CE1

**Comment by saudet** [comment]
@edschwarz Can you try setting the OMP_NUM_THREADS=1 environment variable?

**Comment by edschwarz** [comment]
Sorry so long to respond. Yes, this helped - setting the OMP_NUM_THREADS=1 environment variable.

I began to encounter the actual deadlock. Still using 0.9.1 on the same Macbook. No issues on a circa 2014 Mac mini, same OS version.

**Comment by AlexDBlack** [comment]
Closing this as part of migration to mono-repository.
If this is still an issue with 1.0.0-beta, please re-open an issue here: https://github.com/deeplearning4j/deeplearning4j
